tests/virt_autotest/validate_system_health.pm

```diff
@@ -0,0 +1,84 @@
+
+use base "virt_feature_test_base";
+use strict;
+use warnings;
+use testapi;
+use utils;
+use ipmi_backend_utils qw(reconnect_when_ssh_console_broken);
+use Utils::Architectures;
+use virt_autotest::common;
+use virt_autotest::utils qw(is_xen_host start_guests check_port_state check_host_health check_guest_health);
+use virt_utils qw(collect_host_and_guest_logs);
+
+sub prepare_run_test {
+    reconnect_when_ssh_console_broken unless defined(script_run("rm -f /root/{commands_history,commands_failure}", die_on_timeout => 0));
+    script_run("history -c");
+}
+
+sub run_test {
+
+    my $self = shift;
+    my %health_status = ();
+    $health_status{host} = check_host_health;
+    foreach my $guest (keys %virt_autotest::common::guests) {
+        script_run("virsh start $guest", die_on_timeout => 0);
+        if (script_retry("nmap $guest -PN -p ssh | grep open", delay => 2, retry => 60, die => 0) == 0) {
+            $health_status{$guest} = check_guest_health($guest);
+            if ($health_status{$guest} eq 'fail' || get_var('FORCE_UPLOAD_LOGS')) {
+                script_run("ssh root\@$guest 'dmesg > dmesg.txt 2>&1'", die_on_timeout => 0);
+            }
+        }
+        else {
+            $health_status{$guest} = "Unreachable";
+        }
+        record_info("Failed guest", "$guest: $health_status{$guest}", result => 'fail') unless $health_status{$guest} eq 'pass';
+    }
+    my @unhealthy_list = grep { $health_status{$_} ne 'pass' } keys %health_status;
+
+    # Upload logs on x86_64
+    if (is_x86_64 && (get_var('FORCE_UPLOAD_LOGS') || @unhealthy_list != 0)) {
+        script_run("tar cvzf libvirtd.tar.gz /var/log/libvirt", die_on_timeout => 0);
+        if (is_xen_host) {
+            script_run("tar cvzf xen_logs.tar.gz /var/log/xen", die_on_timeout => 0);
+            script_run("xl dmesg > xl_dmesg.txt", die_on_timeout => 0);
+        }
+        script_run("dmesg > dmesg_host.txt", die_on_timeout => 0);
+        my $host_extra_logs = "/root/libvirtd.tar.gz /root/dmesg_host.txt /root/xen_logs.tar.gz xl_dmesg.txt";
+        my $guest_extra_logs = "/root/dmesg.txt";
+        if (get_var('FORCE_UPLOAD_LOGS')) {
+            collect_host_and_guest_logs(join(' ', grep { $_ ne 'host' } keys %health_status), $host_extra_logs, $guest_extra_logs);
+        }
+        else {
+            collect_host_and_guest_logs(join(' ', grep { $_ ne 'host' } @unhealthy_list), $host_extra_logs, $guest_extra_logs);
+        }
+        $self->upload_coredumps;
+        save_screenshot;
+    }
+    die "Host or guests are not healthy!" unless @unhealthy_list == 0;
+}
+
+sub post_fail_hook {
+    my $self = shift;
+    diag("Module validate_system_health post fail hook starts.");
+    $self->junit_log_provision((caller(0))[3]);
+    reconnect_when_ssh_console_broken unless defined(script_run("history -w /root/commands_history", die_on_timeout => 0));
+    $self->upload_coredumps;
+    collect_host_and_guest_logs;
```

Comment:
As I said before, your logic is confusing. If you already FORCE_UPLOAD_LOGS, you do not need to do collect_host_and_guest_logs again. Could you help refine your logic a little bit ?

Reply:
`FORCE_UPLOAD_LOGS` is to collect_host_and_guest_logs in validate_system_health. It is not used elsewhere.

Reply:
The uploaded logs can be a little duplicated on specific conditions. Is it possible to avoid this to some extent, eg by more condition checks on variables like FORCE_UPLOAD_LOGS?

Reply:
what duplication do you mean? `FORCE_UPLOAD_LOGS` take effects only in validate_system_health test module for a passed test. No logs would be uploaded for a successful test without it.

Reply:
Oh, you and Wayne talked the log collection in post_fail_hook. I just removed it.